File: server/lib/repositories/product_repository.dart
import '../db_connection.dart';

class ProductRepository {
  Future<List<Map<String, dynamic>>> findAll() async {
    final pool = await DBConnection.getPool();
    final sql = '''
      SELECT p.*, c.name as category_name 
      FROM products p 
      LEFT JOIN categories c ON p.category_id = c.id 
      ORDER BY p.name
    ''';
    final result = await pool.execute(sql);
    return result.rows.map((row) => row.assoc()).toList();
  }

  Future<List<Map<String, dynamic>>> findByCategory(int categoryId) async {
    final pool = await DBConnection.getPool();
    final sql = '''
      SELECT p.*, c.name as category_name 
      FROM products p 
      LEFT JOIN categories c ON p.category_id = c.id 
      WHERE p.category_id = :catId
      ORDER BY p.name
    ''';
    final result = await pool.execute(sql, {'catId': categoryId});
    return result.rows.map((row) => row.assoc()).toList();
  }

  Future<Map<String, dynamic>?> findById(int id) async {
    final pool = await DBConnection.getPool();
    final sql = '''
      SELECT p.*, c.name as category_name 
      FROM products p 
      LEFT JOIN categories c ON p.category_id = c.id 
      WHERE p.id = :id
    ''';
    final result = await pool.execute(sql, {'id': id});
    if (result.rows.isNotEmpty) {
      return result.rows.first.assoc();
    }
    return null;
  }

  Future<int> save(Map<String, dynamic> product) async {
    final pool = await DBConnection.getPool();
    final sql = '''
      INSERT INTO products (sku, name, price, stock, category_id, discount_percentage) 
      VALUES (:sku, :name, :price, :stock, :categoryId, :discount)
    ''';
    final result = await pool.execute(sql, {
      'sku': product['sku'],
      'name': product['name'],
      'price': product['price'],
      'stock': product['stock'],
      'categoryId': product['category_id'],
      'discount': product['discount_percentage'],
    });
    return result.lastInsertID.toInt();
  }

  Future<void> update(int id, Map<String, dynamic> product) async {
    final pool = await DBConnection.getPool();
    final sql = '''
      UPDATE products 
      SET sku=:sku, name=:name, price=:price, stock=:stock, category_id=:categoryId, discount_percentage=:discount 
      WHERE id=:id
    ''';
    await pool.execute(sql, {
      'sku': product['sku'],
      'name': product['name'],
      'price': product['price'],
      'stock': product['stock'],
      'categoryId': product['category_id'],
      'discount': product['discount_percentage'],
      'id': id,
    });
  }

  Future<void> updateStock(int productId, int quantity) async {
    final pool = await DBConnection.getPool();
    await pool.execute(
      'UPDATE products SET stock = stock - :qty WHERE id = :id',
      {'qty': quantity, 'id': productId},
    );
  }

  Future<void> deleteById(int id) async {
    final pool = await DBConnection.getPool();
    await pool.execute('DELETE FROM products WHERE id = :id', {'id': id});
  }
}

File: src/main/java/com/mycompany/repository/ProductRepository.java
package com.mycompany.repository;

import com.mycompany.model.Product;
import com.mycompany.db.DBConnection;
import java.sql.*;
import java.util.ArrayList;
import java.util.List;

/**
 * ProductRepository - Repository Pattern
 * Memisahkan logika akses data dari business logic
 */
public class ProductRepository {
    
    /**
     * Find product by ID
     */
    public Product findById(int id) throws SQLException {
        String sql = "SELECT p.*, c.name as category_name FROM products p " +
                     "LEFT JOIN categories c ON p.category_id = c.id " +
                     "WHERE p.id = ?";
        
        try (Connection conn = DBConnection.getConnection();
             PreparedStatement ps = conn.prepareStatement(sql)) {
            
            ps.setInt(1, id);
            ResultSet rs = ps.executeQuery();
            
            if (rs.next()) {
                return mapResultSetToProduct(rs);
            }
            return null;
        }
    }
    
    /**
     * Find all products
     */
    public List<Product> findAll() throws SQLException {
        List<Product> products = new ArrayList<>();
        String sql = "SELECT p.*, c.name as category_name FROM products p " +
                     "LEFT JOIN categories c ON p.category_id = c.id " +
                     "ORDER BY p.name";
        
        try (Connection conn = DBConnection.getConnection();
             PreparedStatement ps = conn.prepareStatement(sql)) {
            
            ResultSet rs = ps.executeQuery();
            while (rs.next()) {
                products.add(mapResultSetToProduct(rs));
            }
        }
        return products;
    }
    
    /**
     * Find products by category
     */
    public List<Product> findByCategory(int categoryId) throws SQLException {
        List<Product> products = new ArrayList<>();
        String sql = "SELECT p.*, c.name as category_name FROM products p " +
                     "LEFT JOIN categories c ON p.category_id = c.id " +
                     "WHERE p.category_id = ? " +
                     "ORDER BY p.name";
        
        try (Connection conn = DBConnection.getConnection();
             PreparedStatement ps = conn.prepareStatement(sql)) {
            
            ps.setInt(1, categoryId);
            ResultSet rs = ps.executeQuery();
            
            while (rs.next()) {
                products.add(mapResultSetToProduct(rs));
            }
        }
        return products;
    }
    
    /**
     * Save new product
     */
    public void save(Product product) throws SQLException {
        String sql = "INSERT INTO products (sku, name, price, stock, category_id, discount_percentage) " +
                     "VALUES (?, ?, ?, ?, ?, ?)";
        
        try (Connection conn = DBConnection.getConnection();
             PreparedStatement ps = conn.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS)) {
            
            ps.setString(1, product.getSku());
            ps.setString(2, product.getName());
            ps.setBigDecimal(3, product.getPrice());
            ps.setInt(4, product.getStock());
            ps.setInt(5, product.getCategory().getId());
            ps.setInt(6, product.getDiscountPercentage());
            
            ps.executeUpdate();
            
            ResultSet rs = ps.getGeneratedKeys();
            if (rs.next()) {
                product.setId(rs.getInt(1));
            }
        }
    }
    
    /**
     * Update existing product
     */
    public void update(Product product) throws SQLException {
        String sql = "UPDATE products SET sku=?, name=?, price=?, stock=?, category_id=?, discount_percentage=? " +
                     "WHERE id=?";
        
        try (Connection conn = DBConnection.getConnection();
             PreparedStatement ps = conn.prepareStatement(sql)) {
            
            ps.setString(1, product.getSku());
            ps.setString(2, product.getName());
            ps.setBigDecimal(3, product.getPrice());
            ps.setInt(4, product.getStock());
            ps.setInt(5, product.getCategory().getId());
            ps.setInt(6, product.getDiscountPercentage());
            ps.setInt(7, product.getId());
            
            ps.executeUpdate();
        }
    }
    
    /**
     * Update product stock (used after sale)
     */
    public void updateStock(int productId, int quantity) throws SQLException {
        String sql = "UPDATE products SET stock = stock - ? WHERE id = ?";
        
        try (Connection conn = DBConnection.getConnection();
             PreparedStatement ps = conn.prepareStatement(sql)) {
            
            ps.setInt(1, quantity);
            ps.setInt(2, productId);
            ps.executeUpdate();
        }
    }
    
    /**
     * Delete product by ID
     */
    public void deleteById(int id) throws SQLException {
        String sql = "DELETE FROM products WHERE id = ?";
        
        try (Connection conn = DBConnection.getConnection();
             PreparedStatement ps = conn.prepareStatement(sql)) {
            
            ps.setInt(1, id);
            ps.executeUpdate();
        }
    }
    
    /**
     * Map ResultSet to Product object
     */
    private Product mapResultSetToProduct(ResultSet rs) throws SQLException {
        Product product = new Product();
        product.setId(rs.getInt("id"));
        product.setSku(rs.getString("sku"));
        product.setName(rs.getString("name"));
        product.setPrice(rs.getBigDecimal("price"));
        product.setStock(rs.getInt("stock"));
        // Diskon per produk (0-100)
        product.setDiscountPercentage(rs.getInt("discount_percentage"));
        
        // Create category object
        com.mycompany.model.Category category = new com.mycompany.model.Category();
        category.setId(rs.getInt("category_id"));
        category.setName(rs.getString("category_name"));
        product.setCategory(category);
        
        return product;
    }
}
